packages/simple-cache-provider/src/SimpleCacheProvider.js

```diff
@@ -0,0 +1,201 @@
+}
+
+export function createCache(invalidator: () => mixed): Cache {
+  const resourceCache: ResourceCache = new Map();
+
+  function getRecord<K, V>(resourceType: any, key: K): Record<V> {
+    if (__DEV__) {
+      warning(
+        typeof resourceType !== 'string' && typeof resourceType !== 'number',
+        'Invalid resourceType: Expected a symbol, object, or function, but ' +
+          'instead received: %s. Strings and numbers are not permitted as ' +
+          'resource types.',
+        resourceType,
+      );
+    }
+
+    let recordCache = resourceCache.get(resourceType);
+    if (recordCache !== undefined) {
+      const record = recordCache.get(key);
+      if (record !== undefined) {
+        return record;
+      }
+    } else {
+      recordCache = new Map();
+      resourceCache.set(resourceType, recordCache);
+    }
+
+    const record = {
+      status: Empty,
+      suspender: null,
+      value: null,
+      error: null,
+    };
+    recordCache.set(key, record);
+    return record;
+  }
+
+  function load<K, V>(record: Record<V>, key: K, miss: K => Promise<V>) {
+    const suspender = miss(key);
+    record.suspender = suspender;
+    suspender.then(
+      value => {
+        // Resource loaded successfully.
+        record.suspender = null;
+        record.status = Resolved;
+        record.value = value;
+      },
+      error => {
+        // Resource failed to load. Stash the error for later so we can throw it
+        // the next time it's requested.
+        record.suspender = null;
+        record.status = Rejected;
+        record.error = error;
+      },
+    );
+    return suspender;
+  }
+
+  const cache: Cache = {
+    invalidate() {
+      invalidator();
+    },
+    preload<K, V>(resourceType: any, key: K, miss: K => Promise<V>): void {
+      const record: Record<V> = getRecord(resourceType, key);
+      const suspender = record.suspender;
+      if (suspender !== null) {
+        // There's already a pending request.
+        return;
+      }
+      switch (record.status) {
+        case Empty:
+          // Warm the cache.
+          load(record, key, miss);
+          return;
+        case Resolved:
+        case Rejected:
+          // The resource is already in the cache.
+          return;
+      }
+    },
+    read<K, V>(resourceType: any, key: K, miss: K => Promise<V>): V {
+      const record: Record<V> = getRecord(resourceType, key);
+      const suspender = record.suspender;
+      if (suspender !== null) {
+        // There's already a pending request.
+        throw suspender;
+      }
+      switch (record.status) {
+        case Empty:
+          // Load the requested resource.
+          throw load(record, key, miss);
+        case Resolved:
+          return ((record.value: any): V);
+        case Rejected:
+        default:
+          // The requested resource previously failed loading.
+          // TODO: When should we try again?
+          throw record.error;
```

Comment:
Ideally we should keep the error state and let the user display the message with a “try again” button. Is this what it does?

Reply:
Sort of. The error is cached so the next time that resource is read, the cache will throw the error so the user can capture it with an error boundary. But I don’t know when we should reset it to empty. I guess right here, right before we throw? But what if multiple components attempt to read the same value? They should all throw, not just the first one. Maybe we should reset it after n milliseconds.